usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -872,14 +878,16 @@ libzfs_mnttab_add(libzfs_handle_t *hdl, const char *special,
 {
 	mnttab_node_t *mtn;
 
-	if (avl_numnodes(&hdl->libzfs_mnttab_cache) == 0)
-		return;
-	mtn = zfs_alloc(hdl, sizeof (mnttab_node_t));
-	mtn->mtn_mt.mnt_special = zfs_strdup(hdl, special);
-	mtn->mtn_mt.mnt_mountp = zfs_strdup(hdl, mountp);
-	mtn->mtn_mt.mnt_fstype = zfs_strdup(hdl, MNTTYPE_ZFS);
-	mtn->mtn_mt.mnt_mntopts = zfs_strdup(hdl, mntopts);
-	avl_add(&hdl->libzfs_mnttab_cache, mtn);
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
```

Comment:
Use mutex_enter(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock.